typing/ctype.ml

```diff
@@ -2835,15 +2835,17 @@ and unify3 uenv t1 t1' t2 t2' =
     end;
     try
       begin match (d1, d2) with
-        (Tarrow (l1, t1, u1, c1), Tarrow (l2, t2, u2, c2))
-        when compatible_labels ~in_pattern_mode:(in_pattern_mode uenv) l1 l2 ->
-          unify uenv t1 t2; unify uenv u1 u2;
-          begin match is_commu_ok c1, is_commu_ok c2 with
-          | false, true -> set_commu_ok c1
-          | true, false -> set_commu_ok c2
-          | false, false -> link_commu ~inside:c1 c2
-          | true, true -> ()
-          end
+        (Tarrow (l1, t1, u1, c1), Tarrow (l2, t2, u2, c2)) ->
+          if
+            compatible_labels ~in_pattern_mode:(in_pattern_mode uenv) l1 l2
+          then begin unify uenv t1 t2; unify uenv u1 u2;
+            match is_commu_ok c1, is_commu_ok c2 with
+            | false, true -> set_commu_ok c1
+            | true, false -> set_commu_ok c2
+            | false, false -> link_commu ~inside:c1 c2
+            | true, true -> ()
+          end else
+            raise_for Unify (Function_label_mismatch {got=l1; expected=l2})
```

Comment:
Nitpick: wouldn't this be easier on the eye with the error case first?

```ocaml
if not (compatible_labels ...) then
  raise_for ...;
unify uenv t1 t2; unify uenv u1 u2;
begin match ... with
  ...
end
```

(Same question in the other, simpler cases.)

Reply:
Why not?

Reply:
Woudn't it be even simpler if we moved the raise inside the `compatible_labels` function ?
This would reduce code duplication (and even more if more label comparison are added).

Reply:
There are calls to `compatible_labels` in `mcomp` and the subtyping mode that doesn't raise the exception, either because we are only checking compatibility or because we are storing an error trace for later use.

However, we could have a `raise_on_incompatible_labels` function which has the advantage of keeping the control flow obvious.